autonat.go

```diff
@@ -75,198 +70,222 @@ func NewAutoNAT(ctx context.Context, h host.Host, getAddrs GetAddrs) AutoNAT {
+	status := as.status.Load().(autoNATResult)
+	as.emitChange.Emit(event.EvtLocalReachabilityChanged{Reachability: status.Reachability})
 }
 
+// PublicAddr returns the publicly connectable Multiaddr of this node if one is known.
 func (as *AmbientAutoNAT) PublicAddr() (ma.Multiaddr, error) {
-	as.mx.Lock()
-	defer as.mx.Unlock()
-
-	if as.status != NATStatusPublic {
+	s := as.status.Load().(autoNATResult)
+	if s.Reachability != network.ReachabilityPublic {
 		return nil, errors.New("NAT Status is not public")
 	}
 
-	return as.addr, nil
+	return s.address, nil
 }
 
 func (as *AmbientAutoNAT) background() {
 	// wait a bit for the node to come online and establish some connections
 	// before starting autodetection
-	select {
-	case <-time.After(AutoNATBootDelay):
-	case <-as.ctx.Done():
-		return
-	}
+	delay := AutoNATBootDelay
+
+	netChangeChan := as.subAddrChange.Out()
 
 	for {
-		as.autodetect()
+		select {
+		// new connection occured.
+		case conn := <-as.candidatePeers:
+			if conn.Stat().Direction == network.DirInbound && manet.IsPublicAddr(conn.RemoteMultiaddr()) {
+				as.lastInbound = time.Now()
+			}
 
-		delay := AutoNATRefreshInterval
-		if as.status == NATStatusUnknown {
-			delay = AutoNATRetryInterval
-		}
+		case <-netChangeChan:
+			if as.confidence > 1 {
+				as.confidence--
+			}
 
-		select {
+		// probe finished.
+		case result, ok := <-as.observations:
+			if !ok {
+				return
+			}
+			as.recordObservation(result)
 		case <-time.After(delay):
 		case <-as.ctx.Done():
 			return
 		}
-	}
-}
-
-func (as *AmbientAutoNAT) autodetect() {
-	peers := as.getPeers()
-
-	if len(peers) == 0 {
-		log.Debugf("skipping NAT auto detection; no autonat peers")
-		return
-	}
-
-	cli := NewAutoNATClient(as.host, as.getAddrs)
-	ctx, cancel := context.WithTimeout(as.ctx, AutoNATRequestTimeout)
-	defer cancel()
 
-	var result struct {
-		sync.Mutex
-		private int
-		public  int
-		pubaddr ma.Multiaddr
+		delay = as.scheduleProbe()
 	}
+}
 
-	probe := 3 - as.confidence
-	if probe == 0 {
-		probe = 1
-	}
-	if probe > len(peers) {
-		probe = len(peers)
+// scheduleProbe calculates when the next probe should be scheduled for,
+// and launches it if that time is now.
+func (as *AmbientAutoNAT) scheduleProbe() time.Duration {
+	// Our baseline is a probe every 'AutoNATRefreshInterval'
+	// This is modulated by:
+	// * recent inbound connections make us willing to wait up to 2x longer between probes.
+	// * low confidence makes us speed up between probes.
+	fixedNow := time.Now()
+	currentStatus := as.status.Load().(autoNATResult)
+
+	nextProbe := fixedNow
+	if !as.lastProbe.IsZero() {
+		untilNext := AutoNATRefreshInterval
+		if currentStatus.Reachability == network.ReachabilityUnknown {
+			untilNext = AutoNATRetryInterval
+		} else if currentStatus.Reachability == network.ReachabilityPublic && as.lastInbound.After(as.lastProbe) {
+			untilNext *= 2
+		} else if as.confidence < 3 {
+			untilNext = AutoNATRetryInterval
```

Comment:
@willscott 

Consider the following sequence of events:

1. We do a probe which changes our current status to `ReachabilityPublic`.
2. We get an inbound connection and update our `lastInbound` to be after `lastProbe`.
3. The peer changes it' s set of listen addresses and so we reduce our confidence after reading the `EvtLocalAddressChanged` event.

The current code would delay the probe because of the inbound connection but would totally discount the fact that our address has changed (which should speed up the probe) _AFTER_ the inbound connection. I think the  confidence based branch should come BEFORE the `lastInbound` based branch or maybe we can find a way to consider both pieces of information with a scaling factor & come up with a balanced value of `untilNext` ?

Reply:
good catch. changed order